### PR TITLE
Add glusterd2's PID in peer list response

### PR DESCRIPTION
**Pull request by Madhu-1**
when glusterd2 starts it adds its PID to ETCD store.
added peer PID to peer list and peer status response

Fixes:#899
Signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@prashanthpai @aravindavk  please review

**Comment by Madhu-1** [comment]
Sample CLI output

```
./build/glustercli --glusterd-host="10.70.42.235" peer list
+--------------------------------------+-----------------------------------+--------------------+--------------------+--------+-------+
|                  ID                  |               NAME                |  CLIENT ADDRESSES  |   PEER ADDRESSES   | ONLINE |  PID  |
+--------------------------------------+-----------------------------------+--------------------+--------------------+--------+-------+
| 4f2d83c4-eaeb-4b01-87fe-b0383b9502ab | dhcp43-75.lab.eng.blr.redhat.com  | 127.0.0.1:24007    | 10.70.43.75:24008  | yes    |  7447 |
|                                      |                                   | 10.70.43.75:24007  |                    |        |       |
| 6de1d304-a82d-41d1-8919-223dbc4ba2cc | dhcp43-235.lab.eng.blr.redhat.com | 127.0.0.1:24007    | 10.70.43.235:24008 | yes    | 23171 |
|                                      |                                   | 10.70.43.235:24007 |                    |        |       |
| ee95085f-fef7-48c7-982b-8e51b05cee57 | dhcp42-235.lab.eng.blr.redhat.com | 127.0.0.1:24007    | 10.70.42.235:24008 | yes    |  2755 |
|                                      |                                   | 10.70.42.235:24007 |                    |        |       |
+--------------------------------------+-----------------------------------+--------------------+--------------------+--------+-------+

```


```
./build/glustercli --glusterd-host="10.70.42.235" peer list
+--------------------------------------+-----------------------------------+--------------------+----------------------------------------+--------+------+
|                  ID                  |               NAME                |  CLIENT ADDRESSES  |             PEER ADDRESSES             | ONLINE | PID  |
+--------------------------------------+-----------------------------------+--------------------+----------------------------------------+--------+------+
| 10fe1225-c338-4506-8657-b36ee9337a2e | dhcp43-75.lab.eng.blr.redhat.com  | 127.0.0.1:24007    | dhcp43-75.lab.eng.blr.redhat.com:24008 | no     |      |
|                                      |                                   | 10.70.43.75:24007  | 10.70.43.75:24008                      |        |      |
| 82e22ef6-96f4-4e87-ace7-d59e8eca69cc | dhcp42-235.lab.eng.blr.redhat.com | 127.0.0.1:24007    | 10.70.42.235:24008                     | yes    | 2551 |
|                                      |                                   | 10.70.42.235:24007 |                                        |        |      |
| e9b77dba-71e9-46fc-afcb-06f77da3c064 | dhcp43-235.lab.eng.blr.redhat.com | 127.0.0.1:24007    | 10.70.43.235:24008                     | no     |      |
|                                      |                                   | 10.70.43.235:24007 |                                        |        |      |
+--------------------------------------+-----------------------------------+--------------------+----------------------------------------+--------+------+

````

File: plugins/smartvol/bricksplanner/utils.go [hunk 1]
```diff
@@ -42,7 +42,8 @@ func getAvailableVgs(req *smartvolapi.Volume) ([]Vg, error) {
 
 	for _, p := range peers {
 		// If Peer is not online, do not consider this device/peer
-		if !store.Store.IsNodeAlive(p.ID) {
+		_, online := store.Store.IsNodeAlive(p.ID)
```
Comment (prashanthpai) [inline plugins/smartvol/bricksplanner/utils.go]: Suggestion: You can use the short notation as that variable isn't used elsewhere 

```go
if _, online := store.Store.IsNodeAlive(p.ID); !online {
    ...
}
```

File: pkg/api/peer_req_resp.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ type Peer struct {
 	PeerAddresses   []string          `json:"peer-addresses"`
 	ClientAddresses []string          `json:"client-addresses"`
 	Online          bool              `json:"online"`
+	PID             string            `json:"pid,omitempty"`
```
Comment (prashanthpai) [inline pkg/api/peer_req_resp.go]: This should be an integer. Displaying it in CLI as not zero is a different problem.

File: glusterd2/transaction/transaction.go [hunk 3]
```diff
@@ -101,7 +101,8 @@ func (t *Txn) checkAlive() error {
 
 	for _, node := range t.Nodes {
 		// TODO: Using prefixed query, get all alive nodes in a single etcd query
-		if !store.Store.IsNodeAlive(node) {
+		_, online := store.Store.IsNodeAlive(node)
```
Comment (prashanthpai) [inline glusterd2/transaction/transaction.go]: Suggestion: You can use the short notation as that variable isn't used elsewhere 

```go
if _, online := store.Store.IsNodeAlive(p.ID); !online {
    ...
}
```

File: glusterd2/store/liveness.go [hunk 4]
```diff
@@ -16,8 +19,12 @@ const (
 	LivenessKeyPrefix = "alive/"
 )
 
-// IsNodeAlive returns true if the node specified is alive as seen by the store
-func (s *GDStore) IsNodeAlive(nodeID interface{}) bool {
+type gPID struct {
+	PID int
+}
+
+// IsNodeAlive returns true and pid if the node specified is alive as seen by the store
```
Comment (prashanthpai) [inline glusterd2/store/liveness.go]: Make this return an int. Displaying in CLI is a different problem.

File: glusterd2/store/liveness.go [hunk 5]
```diff
@@ -39,10 +46,17 @@ func (s *GDStore) IsNodeAlive(nodeID interface{}) bool {
...
 	resp, err := s.Get(ctx, key)
 	if err != nil {
-		return false
+		return "", false
 	}
 
-	return resp.Count == 1
+	if resp.Count == 1 {
+		var p gPID
+		if err := json.Unmarshal(resp.Kvs[0].Value, &p); err != nil {
```
Comment (prashanthpai) [inline glusterd2/store/liveness.go]: You don't need json here. Convert bytes to string to int.

File: glusterd2/store/liveness.go [hunk 6]
```diff
@@ -51,7 +65,9 @@ func (s *GDStore) publishLiveness() error {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 
-	_, err := s.Put(ctx, key, "", clientv3.WithLease(s.Session.Lease()))
+	p := gPID{PID: os.Getpid()}
+	data, _ := json.Marshal(p)
```
Comment (prashanthpai) [inline glusterd2/store/liveness.go]: No json needed. Convert int to string.

File: glusterd2/store/liveness.go [hunk 7]
```diff
@@ -39,19 +41,26 @@ func (s *GDStore) IsNodeAlive(nodeID interface{}) bool {
 	key := LivenessKeyPrefix + keySuffix
 	resp, err := s.Get(ctx, key)
 	if err != nil {
-		return false
+		return 0, false
 	}
 
-	return resp.Count == 1
+	if resp.Count == 1 {
+		pid := string(resp.Kvs[0].Value)
```
Comment (prashanthpai) [inline glusterd2/store/liveness.go]: The variable name `id` and `pid` indicate that they are two different things. You can simply do this instead:

```go
pid, err := strconv.Atoi(string(resp.Kvs[0].Value))
```

File: glusterd2/store/liveness.go [hunk 8]
```diff
@@ -39,19 +41,26 @@ func (s *GDStore) IsNodeAlive(nodeID interface{}) bool {
...
+		id, err := strconv.Atoi(pid)
+		if err != nil {
```
Comment (prashanthpai) [inline glusterd2/store/liveness.go]: You have to this error as it's very serious one. `Atoi` should not fail.

File: glusterd2/store/liveness.go (hunk 7 above)
Comment (Madhu-1) [reply glusterd2/store/liveness.go]: done

File: glusterd2/store/liveness.go (hunk 8 above)
Comment (Madhu-1) [reply glusterd2/store/liveness.go]: done

**Pull request merged**